toolchain/check/handle_class.cpp

```diff
@@ -105,11 +108,19 @@ static auto BuildClassDecl(Context& context, Parse::AnyClassDeclId parse_node)
 
     // Build the `Self` type.
     auto& class_info = context.classes().Get(class_decl.class_id);
-    class_info.self_type_id = context.GetClassType(class_decl.class_id);
+    auto [self_const_id, self_type_id] =
+        context.GetClassType(class_decl.class_id);
+    class_info.self_type_id = self_type_id;
+
+    // Provide the ClassType for name references.
+    // TODO: This may need to differ for generic types, but should still take
+    // name lookup into account.
+    const_id = self_const_id;
   }
 
   // Write the class ID into the ClassDecl.
-  context.ReplaceInstBeforeConstantUse(class_decl_id, {parse_node, class_decl});
+  context.ReplaceInstBeforeConstantUse(class_decl_id, {parse_node, class_decl},
+                                       const_id);
```

Comment:
I would prefer that we create the class type in `TryEvalInst` rather than bypassing constant evaluation here. `constant_values` is currently a cache for the result of the constant evaluator and I think it's valuable that we keep it that way. We can still have the evaluator pull the constant out of the class's `self_type_id` rather than redoing `ClassType` canonicalization, and the small extra cost per class declaration seems fine to me.

Reply:
I was thinking about this, and how about this approach instead, also dropping GetClassType?

Reply:
Thanks, this is great.